src/test/java/hudson/plugins/clover/DeclarativePipelineTest.java

```diff
@@ -0,0 +1,86 @@
+package hudson.plugins.clover;
+
+import org.jenkinsci.plugins.workflow.cps.CpsFlowDefinition;
+import org.jenkinsci.plugins.workflow.job.WorkflowJob;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.jvnet.hudson.test.BuildWatcher;
+import org.jvnet.hudson.test.JenkinsRule;
+
+public class DeclarativePipelineTest {
+    // BuildWatcher echoes job output to stderr as it arrives
+    @ClassRule
+    public static BuildWatcher buildWatcher = new BuildWatcher();
+
+    @Rule
+    public JenkinsRule r = new JenkinsRule();
+
+    @Test
+    public void smokes() throws Exception {
+        final WorkflowRun run = runPipeline(m(
+                "pipeline {",
+                "  agent any",
+                "  stages {",
+                "    stage('Hello World') {",
+                "      steps {",
+                "        echo('hello' + ' from' + ' echo')",
+                "        writeFile file: 'clover.xml', text: '''",
+                "<coverage generated=\"1183450300322\" clover=\"1.3.13\">",
+                "   <project timestamp=\"1183450141125\" name=\"Maven Cloverreport\">",
+                "      <metrics classes=\"5\" methods=\"10\" conditionals=\"2\" files=\"4\" packages=\"1\" coveredstatements=\"2\" loc=\"137\" ncloc=\"70\" coveredmethods=\"1\" coveredconditionals=\"1\" statements=\"18\" coveredelements=\"4\" elements=\"30\"/>",
+                "      <package name=\"hudson.plugins.clover\">",
+                "         <metrics conditionals=\"2\" methods=\"10\" classes=\"5\" files=\"4\" ncloc=\"70\" coveredstatements=\"2\" coveredmethods=\"1\" coveredconditionals=\"1\" statements=\"18\" loc=\"137\" coveredelements=\"4\" elements=\"30\"/>",
+                "         <file name=\"CloverPublisher.java\">",
+                "            <class name=\"CloverPublisher\">",
+                "               <metrics conditionals=\"0\" methods=\"4\" coveredstatements=\"0\" coveredmethods=\"0\" coveredconditionals=\"0\" statements=\"5\" coveredelements=\"0\" elements=\"9\"/>",
+                "            </class>",
+                "            <class name=\"CloverPublisher.DescriptorImpl\">",
+                "               <metrics conditionals=\"0\" methods=\"4\" coveredstatements=\"0\" coveredmethods=\"0\" coveredconditionals=\"0\" statements=\"6\" coveredelements=\"0\" elements=\"10\"/>",
+                "            </class>",
+                "            <metrics conditionals=\"0\" methods=\"8\" classes=\"2\" ncloc=\"42\" coveredstatements=\"0\" coveredmethods=\"0\" coveredconditionals=\"0\" statements=\"11\" loc=\"79\" coveredelements=\"0\" elements=\"19\"/>",
+                "            <line num=\"27\" count=\"0\" type=\"method\"/>",
+                "         </file>",
+                "      </package>",
+                "   </project>",
+                "</coverage>",
```

Comment:
I recommend moving this to an external file, put it in `src/test/resources` directory.

We could have a helper function to read this resource file into String and next concatenate with the rest.

Reply:
Something like:

```
public static String fromResource(String resourceName) {
        try {
            final URL resource = DeclarativePipelineTest.class.getResource(resourceName);
            if (resource == null) {
                throw new NullPointerException("Resource not found for " + resourceName);
            }
            try (Scanner scanner = new Scanner(resource.openStream(), StandardCharsets.UTF_8.name())) {
                return scanner.useDelimiter("\\A").next();
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
}
```

Reply:
You could even move the entire pipeline script to a separate file. But this is optional.

Reply:
Added in 20ab139990542c4fb3153869f6f891302fe125d9